Add EPMD Exposure Blog (#493)
erlef/website: priv/posts/security/20241212210148_epmd-public-exposure.md

  [https://www.rabbitmq.com/docs/networking#epmd](https://www.rabbitmq.com/docs/networking#epmd)  
- The Erlang Ecosystem Foundation’s security hardening guide explains what EPMD is, which ports it uses, and how to close them down:  
  [https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/distribution](https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/distribution)

## What is EPMD and How Does it Work?

Erlang was designed from the ground up for concurrency, distribution, and fault tolerance. To support clustering (distributed Erlang nodes connecting and communicating seamlessly), Erlang relies on EPMD. EPMD runs as a small daemon that keeps track of Erlang nodes and their associated ports. When a node starts with `-name` or `-sname`, it registers with EPMD on the host machine’s port 4369.

In a controlled environment—such as an internal network—this works well. EPMD makes it easy for nodes to discover each other, connect, and form distributed clusters.

**Key points:**
- EPMD listens on port 4369 by default.
- It’s assumed that EPMD is only reachable from within a trusted network.

> - Many Erlang/Elixir build tools and release pipelines (like Rebar3, Mix, and Distillery releases) enable EPMD by default.

Author: I'm not quite sure how Mix behaves with the `remote.vm.args` file.

https://github.com/elixir-lang/elixir/blob/fc2d8d94c0e828f01e0b112efeb0645f99f8d111/lib/mix/lib/mix/tasks/release.init.ex#L56

Please fact-check :)

> - Many Erlang/Elixir build tools and release pipelines (like Rebar3, Mix, and Distillery releases) enable EPMD by default.

Author: Agreed

I see two possibilities about the defaults:

* Erlang by default only starts listening on localhost. Users will then either have to specify `inet_dist_use_interface` with `0.0.0.0` for all interfaces or list specific ones.
* The tools by default don't include `-name` / `-sname` or include ` -kernel inet_dist_use_interface '{127, 0, 0, 1}'`

I personally would prefer to change the default interfaces in Erlang, which will always require a conscious decision by the user.

Also: Not sure if the Mix default is actually problematic with the `remote.vm.args`. I'm not sure how that actually works and will need to investigate. https://github.com/erlef/website/pull/493#discussion_r1882937503

> - Many Erlang/Elixir build tools and release pipelines (like Rebar3, Mix, and Distillery releases) enable EPMD by default.

Reviewer: > I personally would prefer to change the default interfaces in Erlang, which will always require a conscious decision by the user.

By doing that change, you will break every guide/book/video that exists today showing how to setup Erlang Distribution and also all Erlang systems that use distribution correctly. All for a security issue that really should be solved by fixing the firewall to be secure. It would also not fix the RabbitMQ/ejabbard nodes, as the vast majority of such deployments would need Erlang Distribution and EPMD for fault-tolerance.

Also I would again like to point out that Erlang is safe by default, as it does not enable distribution. It is a concious decision by tools and users to enable it. If you enable something called Distributed Erlang, can you really expect that to not listen to external interfaces?

IMO relx, mix release and others should either limit both epmd and dist interface to localhost by default, or just disable it completely and run an [Erlang ssh daemon](https://www.erlang.org/doc/apps/ssh/using_ssh.html#running-an-erlang-ssh-daemon) as part of the release that you can connect to for debugging.

## Why is Exposing EPMD a Bad Idea?

### Risk of Unauthorized Cluster Access

EPMD itself does not enforce strong authentication. Erlang clustering uses a “cookie” as a shared secret, but this cookie is not intended as a robust security mechanism—it’s more of a sanity check to prevent accidental cross-node connections. If an attacker can guess or brute-force the cookie, they could join your cluster.

Once inside the cluster, the attacker could potentially run arbitrary Erlang Remote Procedure Calls (RPCs), giving them full control of the application and underlying system. While no widespread internet-based attacks on EPMD have been documented publicly, the theoretical risk is real. A known brute-force approach is described here:  

> [https://insinuator.net/2017/10/erlang-distribution-rce-and-a-cookie-bruteforcer/](https://insinuator.net/2017/10/erlang-distribution-rce-and-a-cookie-bruteforcer/)

Author: Should we point to here as well? https://github.com/gteissier/erl-matter?tab=readme-ov-file

Reviewer: Yes I think that is a good idea.

### Scope of the Problem

Shodan (a search engine for internet-connected devices) reveals over 85,000 publicly accessible EPMD instances. Among them, around 40,000 are associated with RabbitMQ.  

> Top countries hosting exposed EPMD endpoints include China (18,444), the United States (15,771), and Germany (10,263). Major cloud hosting platforms (DigitalOcean, Hetzner, Aliyun, Tencent, OVH, Amazon, and others) also host significant numbers of exposed instances, typically due to default configurations and lack of firewall restrictions.

Reviewer: ```suggestion
Top countries hosting exposed EPMD endpoints include China (18,444), the United States (15,771), and Germany (10,263). Major cloud hosting platforms (DigitalOcean, Hetzner, Aliyun, Tencent, OVH, Amazon, and others) also host significant numbers of exposed instances, typically due to default configurations and lack of firewall restrictions.
```

> default configurations

I assume that you mean the default configuration of these projects, not of the firewall, as I think most hosting platforms has the firewall closed by default?

[![World Map showing number of exposed EPMD instances](/images/posts/epmd-public-exposure/world-map.png) {: .responsive-image}](https://www.shodan.io/search/report?query=product%3A%22Erlang+Port+Mapper+Daemon%22)

### Not Just Erlang Developers at Risk

RabbitMQ, a popular message broker built on Erlang, runs EPMD under the hood. Many RabbitMQ users may not be aware that EPMD exists or that it’s exposed. This lack of awareness can lead to overlooked vulnerabilities, as users trust defaults without applying basic network hygiene.

## How to Secure Your EPMD

**1. Use a Firewall or Restricted Network Interfaces**

If you don’t need external clustering, ensure EPMD isn’t exposed:

> - Bind EPMD to localhost or a private network interface. In `vm.args`, specify:
>   ```bash
>   -kernel inet_dist_use_interface '{127, 0, 0, 1}'
>   ```

Reviewer: ```suggestion
- Bind Erlang Distrubution and EPMD to localhost or a private network interface. In `vm.args`, specify:
  ```bash
  -kernel inet_dist_use_interface '{127, 0, 0, 1}' -env ERL_EPMD_ADDRESS "127.0.0.1"
  ```
```

This fix only makes Erlang Distribution bind to localhost. You also need to make sure that epmd only listens to localhost.

> - Deploy firewalls, security groups, or network access control lists (ACLs) to ensure port 4369 is not reachable from untrusted networks.

Reviewer: As I said above, restricting 4369 is not enough, you also need to restrict the Erlang Distrubution port, which is a random port.

**2. Remove Clustering Options if Not Needed**

> If you’re not clustering multiple Erlang/RabbitMQ nodes, don’t start the node with `-name` or `-sname`. Without these options, your application won’t register with EPMD at all. If you only run a single RabbitMQ node and do not need distribution, consider disabling or limiting it according to the official networking and clustering documentation.

Reviewer: ```suggestion
If you’re not clustering multiple Erlang/RabbitMQ nodes, don’t start the node with `-name` or `-sname`. Without these options, your application won’t register with EPMD at all. If you only run a single RabbitMQ node and do not need distribution, consider disabling or limiting it according to the official networking and clustering documentation.
```

Many systems use `erl_call` or `remsh` for debugging tools, which requires distribution to be running. So even if you are not clustered to other machines, you may will need/want distrubution running for debugging.

**3. Mitigation is Simple**

A small configuration change or a single firewall rule can often resolve the issue. Once you’ve made changes to close or secure EPMD, scan again to confirm:

> ```bash

Reviewer: Just leaving a note again here, closing 4369 is not enough to be safe.

{
  "title": "Exposed EPMD: A Hidden Security Risk for RabbitMQ and the BEAM Ecosystem",
  "authors": ["The Security Working Group"],
  "slug": "epmd-public-exposure",
  "category": "security",
  "tags": ["security", "epmd", "cybersecurity", "rabbitmq"],
  "datetime": "2024-12-12T21:01:48.619713Z"
}
---
EPMD, essential for Erlang and RabbitMQ clustering, is often exposed online—posing hidden security risks and requiring quick mitigation steps.
---

The Erlang Port Mapper Daemon (EPMD) is a built-in component that helps Erlang-based applications (including RabbitMQ) discover each other’s distribution ports for clustering. Although EPMD itself isn’t directly dangerous, its exposure on the public internet often signals that Erlang Distribution ports are also exposed. This creates a serious security risk: if attackers find these distribution ports, they can potentially join your cluster, run arbitrary code, and compromise your systems. Recent scans have revealed over 85,000 instances of publicly accessible EPMD, with roughly half associated with RabbitMQ servers.

> If left unsecured, exposed Erlang Distribution ports let attackers gain a foothold in your system. Fortunately, mitigation steps are straightforward: disable EPMD if you’re not clustering, or restrict it behind a firewall and proper network configuration—and ensure Erlang Distribution is never exposed to untrusted networks.

Author: ```suggestion
If left unsecured, exposed Erlang Distribution ports let attackers gain a foothold in your system. Fortunately, mitigation steps are straightforward: disable Erlang Distribution if you’re not clustering, or restrict it behind a firewall and proper network configuration—and ensure Erlang Distribution is never exposed to untrusted networks.
```

## TL;DR

**Check if you’re exposed:**
- Run a port scan:  
  ```bash
  nmap -A -sT -p 4369 [Your Public IP]
  ```  
  If the output shows:  
  ```
  PORT     STATE SERVICE VERSION
  4369/tcp open  epmd    Erlang Port Mapper Daemon
  | epmd-info: 
  |   epmd_port: 4369
  |   nodes: 
  |_    rabbit: 25672
  ```
  Your EPMD port is exposed, indicating your Erlang Distribution ports may be discoverable as well.

**For Erlang/Elixir users:**

- **Use a firewall or security group rules to restrict external access**:  
  Close all ports except those explicitly required for your application. Restricting access to only essential services reduces your overall attack surface and helps mitigate various security risks beyond just EPMD exposure.

- **Bind distribution to a non-public interface**:  
  In `vm.args`:
  ```bash

>   -kernel inet_dist_use_interface '{127, 0, 0, 1}'

Reviewer: ```suggestion
  -kernel inet_dist_use_interface '{127, 0, 0, 1}' -env ERL_EPMD_ADDRESS="127.0.0.1"
```

  ```
  Ensuring that distribution ports are only accessible via localhost or a private network interface further reduces the attack surface.

- **Remove or avoid using `-name` or `-sname` if you don’t need clustering**:  

>   If you’re not clustering nodes, don’t assign them a name that registers with EPMD. Without a name, your node won’t appear in EPMD’s registry, making it harder for attackers to locate your distribution ports.

Reviewer: Removing the `-name`/`-sname` disables distribution completely. It is not only removed from EPMD.

[![World Map showing number of exposed EPMD instances](/images/posts/epmd-public-exposure/world-map.png) {: .responsive-image}](https://www.shodan.io/search/report?query=product%3A%22Erlang+Port+Mapper+Daemon%22)

### Not Just Erlang Developers at Risk

RabbitMQ, a popular message broker built on Erlang, runs EPMD and Erlang Distribution under the hood. Many RabbitMQ users may not be aware that these ports are exposed. Without proper network hygiene, this can leave their systems vulnerable, even if they never use Erlang directly.

## How to Secure Your Erlang Distribution (and EPMD)

**1. Use a Firewall or Restricted Network Interfaces**

If you don’t need external clustering, ensure EPMD and the Erlang Distribution ports aren’t exposed:

- Bind Erlang Distrubution and EPMD to localhost or a private network interface. In `vm.args`, specify:
  ```bash
  -kernel inet_dist_use_interface '{127, 0, 0, 1}' -env ERL_EPMD_ADDRESS "127.0.0.1"
  ```

> - Deploy firewalls, security groups, or network access control lists (ACLs) to ensure port 4369 is not reachable from untrusted networks.

Reviewer: Should probably mention here as well that only port 4369 is not enough.
